src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -1643,29 +1754,581 @@ protected Object parseValue(CommandSource source, CommandArgs args) throws Argum
         @Override
         protected Object getValue(String choice) throws IllegalArgumentException {
             UUID uuid = UUID.fromString(choice);
+            boolean found = false;
             for (World world : Sponge.getServer().getWorlds()) {
                 Optional<Entity> ret = world.getEntity(uuid);
                 if (ret.isPresent()) {
-                    return ret.get();
+                    Entity entity = ret.get();
+                    if (this.checkEntity(entity)) {
+                        return entity;
+                    }
+                    found = true;
                 }
             }
+            if (found) {
+                throw new IllegalArgumentException("Input value " + choice + " was not an entity of the required type!");
```

Comment:
Because that has different behaviour, setting the semaphore means the loop continues and may find the correct entity in another world. Though not sure if that's possible because it implies there's an entity with a matching uuid but incorrect type in one world, and matching uuid and correct type in another so...